### PR TITLE
Module 1: Web API Application Development

### PR DESCRIPTION
**Pull request by serhiiyasenev**

### CODE REVIEW

File: src/Api/Controllers/BookingController.cs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            {
+                if (ex.Message.Contains("Not Found"))
+                {
+                    return NotFound(new ProblemDetails { Title = "Not Found by id", Detail = ex.Message });
+                }
+                else if (ex.Message.Contains("already linked"))
+                {
+                    return BadRequest(new ProblemDetails { Title = "Product already linked", Detail = ex.Message });
+                }
+                else
```
Comment (oleksandr-fomenko) [inline src/Api/Controllers/BookingController.cs]: "else" is redundant here

File: src/Api/Controllers/BookingController.cs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+                return CreatedAtAction(nameof(AddBooking), createdBooking);
+            }
+            catch (Exception ex)
+            {
+                if (ex.Message.Contains("Not Found"))
+                {
+                    return NotFound(new ProblemDetails { Title = "Not Found by id", Detail = ex.Message });
+                }
+                else if (ex.Message.Contains("already linked"))
```
Comment (oleksandr-fomenko) [inline src/Api/Controllers/BookingController.cs]: just "if"

File: src/Api/Controllers/BookingController.cs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+                _logger.LogInformation($"Booking email was sent to `{createdBooking.CustomerEmail}`'");
+
+                return CreatedAtAction(nameof(AddBooking), createdBooking);
+            }
+            catch (Exception ex)
+            {
+                if (ex.Message.Contains("Not Found"))
+                {
+                    return NotFound(new ProblemDetails { Title = "Not Found by id", Detail = ex.Message });
```
Comment (oleksandr-fomenko) [inline src/Api/Controllers/BookingController.cs]: provide the id of the item as you did in the GetProductById

File: src/DataAccessLayer/Repositories/BookingDbRepository.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+{
+    public class BookingDbRepository : BaseDbRepository, IBookingRepository
+    {
+        public BookingDbRepository(EfCoreContext efContext) : base(efContext)
+        {
+        }
+
+        public async Task<BookingDto> Add(BookingDto booking, CancellationToken cancellationToken = default)
+        {
+            //booking.Id = Guid.NewGuid();
```
Comment (oleksandr-fomenko) [inline src/DataAccessLayer/Repositories/BookingDbRepository.cs]: please remove unused code

File: src/DataAccessLayer/Repositories/BookingDbRepository.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        public async Task<(IQueryable<BookingDto> FilteredItems, int TotalCount)> GetAll(ItemsRequest request, CancellationToken cancellationToken = default)
+        {
+            var query = _dbContext.Bookings.Include(e => e.Products).AsNoTracking();
+            if (!string.IsNullOrEmpty(request.ItemName))
+            {
+                query = query.Where(item => item.Name.Contains(request.ItemName));
+            }
+            // bottleneck ??
+            int totalCount = await query.CountAsync(cancellationToken);
```
Comment (oleksandr-fomenko) [inline src/DataAccessLayer/Repositories/BookingDbRepository.cs:32]: EntityFramework should translate the CountAsync method into a native query like SELECT COUNT(*) FROM ... which has no problems with the performance even on the huge datasets

File: src/DataAccessLayer/Repositories/ProductDbRepository.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        private readonly EfCoreContext _dbContext;
+
+        public ProductDbRepository(EfCoreContext dbContext)
+        {
+            _dbContext = dbContext;
+        }
+
+        public async Task<ProductDto> Add(ProductDto product, CancellationToken cancellationToken = default)
+        {
+            //product.Id = Guid.NewGuid();
```
Comment (oleksandr-fomenko) [inline src/DataAccessLayer/Repositories/ProductDbRepository.cs]: remove

File: src/DataAccessLayer/Repositories/ProductDbRepository.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        public async Task<int> RemoveItemById(Guid id)
+        {
+            var product = new ProductDto { Id = id };
+            _dbContext.Attach(product);
+            _dbContext.Entry(product).State = EntityState.Deleted;
+            try
+            {
+                return await _dbContext.SaveChangesAsync();
```
Comment (oleksandr-fomenko) [inline src/DataAccessLayer/Repositories/ProductDbRepository.cs]: reason for wrapping it into try-catch?

File: src/InfrastructureLayer/InfrastructureLayer.csproj [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+	<PropertyGroup>
+		<TargetFramework>net7.0</TargetFramework>
+	</PropertyGroup>
+
+	<ItemGroup>
+	  <Compile Remove="Telegram\**" />
+	  <EmbeddedResource Remove="Telegram\**" />
+	  <None Remove="Telegram\**" />
+	</ItemGroup>
```
Comment (oleksandr-fomenko) [inline src/InfrastructureLayer/InfrastructureLayer.csproj]: Looks like you don't need Telegram folder at all

File: src/Api/Controllers/BookingController.cs (hunk 1 above)
Comment (serhiiyasenev) [reply src/Api/Controllers/BookingController.cs]: done

File: src/Api/Controllers/BookingController.cs (hunk 2 above)
Comment (serhiiyasenev) [reply src/Api/Controllers/BookingController.cs]: done

File: src/Api/Controllers/BookingController.cs (hunk 3 above)
Comment (serhiiyasenev) [reply src/Api/Controllers/BookingController.cs]: done (messages created in the service and provided here as `Detail = ex.Message` with Ids)

File: src/DataAccessLayer/Repositories/BookingDbRepository.cs (hunk 4 above)
Comment (serhiiyasenev) [reply src/DataAccessLayer/Repositories/BookingDbRepository.cs]: done

File: src/DataAccessLayer/Repositories/BookingDbRepository.cs (hunk 5 above)
Comment (serhiiyasenev) [reply src/DataAccessLayer/Repositories/BookingDbRepository.cs:32]: thanks!

File: src/DataAccessLayer/Repositories/ProductDbRepository.cs (hunk 6 above)
Comment (serhiiyasenev) [reply src/DataAccessLayer/Repositories/ProductDbRepository.cs]: done

File: src/DataAccessLayer/Repositories/ProductDbRepository.cs (hunk 7 above)
Comment (serhiiyasenev) [reply src/DataAccessLayer/Repositories/ProductDbRepository.cs]: rewrote with `FindAsync` and `Remove`

File: src/InfrastructureLayer/InfrastructureLayer.csproj (hunk 8 above)
Comment (serhiiyasenev) [reply src/InfrastructureLayer/InfrastructureLayer.csproj]: deleted

**Pull request merged**